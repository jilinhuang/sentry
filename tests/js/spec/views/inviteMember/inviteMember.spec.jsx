import React from 'react';
import {shallow, mount} from 'enzyme';
import _ from 'lodash';
import InviteMember from 'app/views/inviteMember/inviteMember';
import {Client} from 'app/api';
import ConfigStore from 'app/stores/configStore';

jest.mock('app/api');

describe('CreateProject', function() {
  beforeEach(function() {
    this.sandbox = sinon.sandbox.create();
    this.sandbox.stub(ConfigStore, 'get').returns({id: 1});
    Client.clearMockResponses();
  });

  afterEach(function() {
    this.sandbox.restore();
  });

  describe('render()', function() {
    const baseProps = {
      params: {
        orgId: 'testOrg'
      }
    };

    const baseContext = {
      context: {
        organization: {
          id: '1',
          slug: 'testOrg',
          teams: [
            {slug: 'bar', id: '1', name: 'bar', hasAccess: true},
            {slug: 'foo', id: '2', name: 'foo', hasAccess: false}
          ]
        },
        router: TestStubs.router(),
        location: {query: {}}
      },
      childContextTypes: {
        organization: React.PropTypes.object,
        location: React.PropTypes.object,
        router: React.PropTypes.object
      }
    };

    it('should render', function() {
      let wrapper = shallow(<InviteMember {...baseProps} />, baseContext);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render no team select when there is only one option', function() {
      let context = _.cloneDeep(baseContext);

      context.context.organization.teams = context.context.organization.teams.slice(0, 1);

      let wrapper = shallow(<InviteMember {...baseProps} />, context);

      expect(wrapper).toMatchSnapshot();
    });

    it('should redirect when no roles available', function() {
      Client.addMockResponse({
        url: '/organizations/testOrg/members/1/',
        body: {
          role_list: [
            {
              role: {
                id: 1,
                name: 'member',
                desc: 'a normal member'
              },
              allowed: false
            }
          ],
          is_invite: true
        }
      });

      let handleSubmitStub = this.sandbox.stub(
        InviteMember.prototype,
        'redirectToMemberPage'
      );
      // üë∫ ‚ö†Ô∏è this is a hack to defeat the method auto binding so we can fully stub the method. It would not be neccessary with es6 class components and it relies on react internals so it's fragile - maxbittker
      const index =
        InviteMember.prototype.__reactAutoBindPairs.indexOf('redirectToMemberPage') + 1;
      InviteMember.prototype.__reactAutoBindPairs[index] = handleSubmitStub;

      let wrapper = mount(<InviteMember {...baseProps} />, baseContext);

      expect(handleSubmitStub.callCount).toEqual(1);
      expect(wrapper.state('loading')).toBe(false);
    });

    it('should render roles when available and allowed, and handle submitting', function(
      done
    ) {
      Client.addMockResponse({
        url: '/organizations/testOrg/members/1/',
        body: {
          role_list: [
            {
              role: {id: 1, name: 'member', desc: 'a normal member'},
              allowed: true
            },
            {
              role: {id: 2, name: 'bar', desc: 'another role'},
              allowed: true
            }
          ],
          is_invite: true
        }
      });

      Client.addMockResponse({
        url: '/organizations/testOrg/members/',
        method: 'POST',
        statusCode: 200,
        body: {}
      });

      let wrapper;

      // üë∫ ‚ö†Ô∏è this is a hack to defeat the method auto binding so we can fully stub the method. It would not be neccessary with es6 class components and it relies on react internals so it's fragile - maxbittker
      const index =
        InviteMember.prototype.__reactAutoBindPairs.indexOf('redirectToMemberPage') + 1;
      InviteMember.prototype.__reactAutoBindPairs[index] = () => {
        expect(wrapper.state('loading')).toBe(true);
        done();
      };

      wrapper = mount(<InviteMember {...baseProps} />, baseContext);

      expect(wrapper.state('loading')).toBe(false);

      let node = wrapper.find('.radio').first();
      node.props().onClick();

      node = wrapper.find('.team-choices > div').first();
      node.props().onClick({preventDefault: () => {}});

      expect(wrapper).toMatchSnapshot();

      node = wrapper.find('.submit-new-team').first();
      node.props().onClick({preventDefault: () => {}});
      expect(wrapper.state('loading')).toBe(false);

      wrapper.setState({email: 'test@email.com'});
      node.props().onClick({preventDefault: () => {}});
    });
  });
});
