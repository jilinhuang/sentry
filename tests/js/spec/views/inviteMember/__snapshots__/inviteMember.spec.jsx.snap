// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateProject render() should render 1`] = `
<HomeContainer>
  <h3>
    Add Member to Organization
  </h3>
  <p>
    You may add a user by their username if they already have an account. Multiple inputs delimited by commas.
  </p>
  <div
    className=""
  >
    <LoadingIndicator
      className="pull-right"
      mini={true}
    />
    <TextField
      disabled={false}
      label="Username(s)"
      name="email"
      onChange={[Function]}
      placeholderlabel="Usernames"
      required={false}
      spellCheck="false"
    />
  </div>
  <div
    className="new-invite-team box"
  >
    <div
      className="box-header"
    >
      <h4>
        Role:
      </h4>
    </div>
    <div
      className="box-content with-padding"
    >
      <ul
        className="radio-inputs"
      />
    </div>
  </div>
  <div
    className="new-invite-team box"
  >
    <div
      className="box-header"
    >
      <h4>
        Team:
      </h4>
    </div>
    <div
      className="grouping-controls team-choices row box-content with-padding"
    >
      <div
        className="col-md-3"
        onClick={[Function]}
      >
        <label
          className="checkbox"
        >
          <Checkbox
            checked={false}
            id="1"
            value="bar"
          />
          bar
          <span
            className="team-slug"
          >
            bar
          </span>
        </label>
      </div>
      <div
        className="col-md-3"
        onClick={[Function]}
      >
        <label
          className="checkbox"
        >
          <Checkbox
            checked={false}
            id="2"
            value="foo"
          />
          foo
          <span
            className="team-slug"
          >
            foo
          </span>
        </label>
      </div>
    </div>
  </div>
  <button
    className="btn btn-primary submit-new-team"
    onClick={[Function]}
  >
    Add Member
  </button>
</HomeContainer>
`;

exports[`CreateProject render() should render no team select when there is only one option 1`] = `
<HomeContainer>
  <h3>
    Add Member to Organization
  </h3>
  <p>
    You may add a user by their username if they already have an account. Multiple inputs delimited by commas.
  </p>
  <div
    className=""
  >
    <LoadingIndicator
      className="pull-right"
      mini={true}
    />
    <TextField
      disabled={false}
      label="Username(s)"
      name="email"
      onChange={[Function]}
      placeholderlabel="Usernames"
      required={false}
      spellCheck="false"
    />
  </div>
  <div
    className="new-invite-team box"
  >
    <div
      className="box-header"
    >
      <h4>
        Role:
      </h4>
    </div>
    <div
      className="box-content with-padding"
    >
      <ul
        className="radio-inputs"
      />
    </div>
  </div>
  <button
    className="btn btn-primary submit-new-team"
    onClick={[Function]}
  >
    Add Member
  </button>
</HomeContainer>
`;

exports[`CreateProject render() should render roles when available and allowed, and handle submitting 1`] = `
<InviteMember
  params={
    Object {
      "orgId": "testOrg",
    }
  }
>
  <HomeContainer>
    <div
      className=" organization-home"
    >
      <div
        className="sub-header flex flex-container flex-vertically-centered"
      >
        <div>
          <ProjectSelector
            organization={
              Object {
                "id": "1",
                "slug": "testOrg",
                "teams": Array [
                  Object {
                    "hasAccess": true,
                    "id": "1",
                    "name": "bar",
                    "slug": "bar",
                  },
                  Object {
                    "hasAccess": false,
                    "id": "2",
                    "name": "foo",
                    "slug": "foo",
                  },
                ],
              }
            }
            projectId={null}
          >
            <div
              className="project-select"
            >
              <h3>
                <Link
                  className="home-crumb"
                  to="/testOrg/"
                >
                  <Link
                    className="home-crumb"
                    onlyActiveOnIndex={false}
                    style={Object {}}
                    to="/testOrg/"
                  >
                    <a
                      className="home-crumb"
                      onClick={[Function]}
                      style={Object {}}
                    >
                      <span
                        className="icon-home"
                      />
                    </a>
                  </Link>
                </Link>
                Select a project
                <DropdownLink
                  anchorRight={false}
                  caret={true}
                  disabled={false}
                  onClose={[Function]}
                  onOpen={[Function]}
                  title=""
                  topLevelClasses="project-dropdown is-empty"
                >
                  <span
                    className="project-dropdown is-empty dropdown"
                  >
                    <a
                      className="dropdown-toggle"
                      data-toggle="dropdown"
                    >
                      <i
                        className="icon-arrow-down"
                      />
                    </a>
                    <ul
                      className="dropdown-menu"
                    >
                      <MenuItem
                        className="empty-projects-item"
                        noAnchor={true}
                      >
                        <li
                          className="empty-projects-item"
                          href={null}
                          role="presentation"
                          title={null}
                        >
                          <div
                            className="empty-message"
                          >
                            You have no projects.
                          </div>
                        </li>
                      </MenuItem>
                    </ul>
                  </span>
                </DropdownLink>
              </h3>
            </div>
          </ProjectSelector>
        </div>
        <div
          className="align-right hidden-xs"
        >
          <Button
            className="tip"
            data-placement="bottom"
            disabled={true}
            priority="primary"
            style={
              Object {
                "marginRight": 5,
              }
            }
            title="You do not have enough permission to create new projects"
          >
            <button
              className="tip button tip button-default button-disabled"
              data-placement="bottom"
              disabled={true}
              onClick={[Function]}
              role="button"
              style={
                Object {
                  "marginRight": 5,
                }
              }
            >
              <FlowLayout
                truncate={false}
              >
                <div
                  className="flow-layout"
                >
                  <span
                    className="button-label"
                  >
                    New Project
                  </span>
                </div>
              </FlowLayout>
            </button>
          </Button>
          <Button
            className="tip"
            data-placement="bottom"
            disabled={true}
            priority="primary"
            title="You do not have enough permission to create new teams"
          >
            <button
              className="tip button tip button-default button-disabled"
              data-placement="bottom"
              disabled={true}
              onClick={[Function]}
              role="button"
            >
              <FlowLayout
                truncate={false}
              >
                <div
                  className="flow-layout"
                >
                  <span
                    className="button-label"
                  >
                    New Team
                  </span>
                </div>
              </FlowLayout>
            </button>
          </Button>
        </div>
      </div>
      <div
        className="container"
      >
        <div
          className="content row"
        >
          <div
            className="col-md-2 org-sidebar"
          >
            <HomeSidebar>
              <div>
                <h6
                  className="nav-header"
                >
                  Organization
                </h6>
                <ul
                  className="nav nav-stacked"
                >
                  <ListLink
                    activeClassName="active"
                    index={false}
                    isActive={[Function]}
                    to="/testOrg/"
                  >
                    <li
                      className=""
                    >
                      <Link
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/testOrg/"
                      >
                        <a
                          onClick={[Function]}
                          style={Object {}}
                        >
                          Dashboard
                        </a>
                      </Link>
                    </li>
                  </ListLink>
                  <ListLink
                    activeClassName="active"
                    index={false}
                    isActive={[Function]}
                    to="/organizations/testOrg/teams/"
                  >
                    <li
                      className=""
                    >
                      <Link
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to="/organizations/testOrg/teams/"
                      >
                        <a
                          onClick={[Function]}
                          style={Object {}}
                        >
                          Projects & Teams
                        </a>
                      </Link>
                    </li>
                  </ListLink>
                </ul>
                <div>
                  <h6
                    className="nav-header with-divider"
                  >
                    Issues
                  </h6>
                  <ul
                    className="nav nav-stacked"
                  >
                    <ListLink
                      activeClassName="active"
                      index={false}
                      to="/organizations/testOrg/issues/assigned/"
                    >
                      <li
                        className=""
                      >
                        <Link
                          onlyActiveOnIndex={false}
                          style={Object {}}
                          to="/organizations/testOrg/issues/assigned/"
                        >
                          <a
                            onClick={[Function]}
                            style={Object {}}
                          >
                            Assigned to Me
                          </a>
                        </Link>
                      </li>
                    </ListLink>
                    <ListLink
                      activeClassName="active"
                      index={false}
                      to="/organizations/testOrg/issues/bookmarks/"
                    >
                      <li
                        className=""
                      >
                        <Link
                          onlyActiveOnIndex={false}
                          style={Object {}}
                          to="/organizations/testOrg/issues/bookmarks/"
                        >
                          <a
                            onClick={[Function]}
                            style={Object {}}
                          >
                            Bookmarks
                          </a>
                        </Link>
                      </li>
                    </ListLink>
                    <ListLink
                      activeClassName="active"
                      index={false}
                      to="/organizations/testOrg/issues/history/"
                    >
                      <li
                        className=""
                      >
                        <Link
                          onlyActiveOnIndex={false}
                          style={Object {}}
                          to="/organizations/testOrg/issues/history/"
                        >
                          <a
                            onClick={[Function]}
                            style={Object {}}
                          >
                            History
                          </a>
                        </Link>
                      </li>
                    </ListLink>
                  </ul>
                </div>
              </div>
            </HomeSidebar>
          </div>
          <div
            className="col-md-10"
          >
            <h3>
              Add Member to Organization
            </h3>
            <p>
              Invite a member to join this organization via their email address. If they do not already have an account, they will first be asked to create one. Multiple emails delimited by commas.
            </p>
            <div
              className=""
            >
              <TextField
                disabled={false}
                label="Email(s)"
                name="email"
                onChange={[Function]}
                placeholderlabel="e.g. teammate@example.com"
                required={false}
                spellCheck="false"
              >
                <div
                  className="control-group"
                >
                  <div
                    className="controls"
                  >
                    <label
                      className="control-label"
                      htmlFor="id-email"
                    >
                      Email(s)
                    </label>
                    <input
                      className="form-control"
                      disabled={false}
                      id="id-email"
                      onChange={[Function]}
                      required={false}
                      spellCheck="false"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </TextField>
            </div>
            <div
              className="new-invite-team box"
            >
              <div
                className="box-header"
              >
                <h4>
                  Role:
                </h4>
              </div>
              <div
                className="box-content with-padding"
              >
                <ul
                  className="radio-inputs"
                >
                  <li
                    className="radio"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    <label
                      style={Object {}}
                    >
                      <Radio
                        checked={true}
                        id={1}
                        readOnly={true}
                        value="member"
                      >
                        <input
                          checked={true}
                          className="radio-select"
                          id={1}
                          readOnly={true}
                          type="radio"
                          value="member"
                        />
                      </Radio>
                      member
                      <div
                        className="help-block"
                      >
                        a normal member
                      </div>
                    </label>
                  </li>
                  <li
                    className="radio"
                    onClick={[Function]}
                    style={Object {}}
                  >
                    <label
                      style={Object {}}
                    >
                      <Radio
                        checked={false}
                        id={2}
                        readOnly={true}
                        value="bar"
                      >
                        <input
                          checked={false}
                          className="radio-select"
                          id={2}
                          readOnly={true}
                          type="radio"
                          value="bar"
                        />
                      </Radio>
                      bar
                      <div
                        className="help-block"
                      >
                        another role
                      </div>
                    </label>
                  </li>
                </ul>
              </div>
            </div>
            <div
              className="new-invite-team box"
            >
              <div
                className="box-header"
              >
                <h4>
                  Team:
                </h4>
              </div>
              <div
                className="grouping-controls team-choices row box-content with-padding"
              >
                <div
                  className="col-md-3"
                  onClick={[Function]}
                >
                  <label
                    className="checkbox"
                  >
                    <Checkbox
                      checked={true}
                      id="1"
                      value="bar"
                    >
                      <input
                        checked={true}
                        className="chk-select"
                        id="1"
                        type="checkbox"
                        value="bar"
                      />
                    </Checkbox>
                    bar
                    <span
                      className="team-slug"
                    >
                      bar
                    </span>
                  </label>
                </div>
                <div
                  className="col-md-3"
                  onClick={[Function]}
                >
                  <label
                    className="checkbox"
                  >
                    <Checkbox
                      checked={false}
                      id="2"
                      value="foo"
                    >
                      <input
                        checked={false}
                        className="chk-select"
                        id="2"
                        type="checkbox"
                        value="foo"
                      />
                    </Checkbox>
                    foo
                    <span
                      className="team-slug"
                    >
                      foo
                    </span>
                  </label>
                </div>
              </div>
            </div>
            <button
              className="btn btn-primary submit-new-team"
              onClick={[Function]}
            >
              Add Member
            </button>
          </div>
        </div>
      </div>
    </div>
  </HomeContainer>
</InviteMember>
`;
